### PR TITLE
Fixed handling of implied index names

### PR DESCRIPTION
**Pull request by georgeaf99**
track #2

### CODE REVIEW

**Comment by xuset** [comment]
With this we won't need the old index file code that didn't work so it can be removed. It's in onFetch in sw.js

**Comment by georgeaf99** [comment]
No, I'm not sure how you could tell the difference between a directory and a file without an extension. Unless metadata is stored about the structure of the directory, then ambiguity will always exist.

**Comment by xuset** [comment]
It is possible to tell whether the request is for a file or directory by looking in the manifest, but doing so isn't entirely necessary. If `/foo` isn't in the manifest then check and see if `/foo/index.html` is in the manifest. The later will is only possible if foo is a directory. Alternatively, to determine if foo is a directory, just look and see if it contains anything, but this only works if the directory is non-empty (also it's not possible to store an empty dir in planktos so edge case averted).

We do need to support /foo and /foo/ though since both are equally valid.

**Comment by georgeaf99** [comment]
Yeah, checking for a /foo file works. This is very similar to the way git treats directories...

**Comment by xuset** [comment]
I tried it out and a request for /foo does work but not in the way that it should. What happens is: the request for /foo gets intercepted, it's not in the manifest so it goes out over http, the server redirects /foo to /foo/, then /foo/ gets intercepted and is in the manifest, so /foo/index.html gets returned.

File: index.js [hunk 1]
```diff
@@ -58,6 +58,23 @@ function getFileBlob (filename) {
 
   return persistent.get(['manifest', 'torrentMeta']).then(result => {
     var [manifest, torrentMeta] = result
+
+    // Try to find an index file if filename is a directory
+    if (filename.endsWith('/')) {
```
Comment (georgeaf99) [inline index.js]: Fixed.

**Comment by georgeaf99** [comment]
Checking to see if `filePath + '/index.html'` is in the manifest won't work, since we have to check if `filePath` ends with a `/` or not. It is also necessary to ensure that the `filePath` is an directory in order to prevent a faulty substitution in this case:
- `/foo/bar`
- `/foo/barindex.html`

I know this is a hide in your bunker from cosmic rays type of bug, but I still think the approach is more correct. I think my code minimally complex in order to capture all of this logic.

Additionally, it is useful to have the ability to default to different types of index files. The current solution provides the ability to default to both `index.html` and `index.htm`. I'm also predicting a use case in the future where a user who has configured their web server to default to `index.php` would want to configure Planktos in the same way. This solution could easily be adapted to accommodate this use case.

**Comment by xuset** [comment]
You're right in that checking if `filePath + '/index.html' won't work because filePath could end in a slash or not, but using `path.join(filePath, 'index.html')` avoids this case by adding/removing a slash only if it's necessary. I don't see how a faulty substitution is possible since `path.join('foo', 'index.html')` will only exist in the manifest if 'foo' is a directory so an explicit check is not necessary.

It's also easy to support multiple index files by appending double pipes. The expression

`foo = a || b || c`

is equivalent to

```
if (a) foo = a
else if (b) foo = b
else foo = c
```

The two approaches seem to cover all cases equally the only difference is the amount of code. Unless I'm missing something?

**Comment by georgeaf99** [comment]
I see what you mean by not explicitly having to do the check. I updated the code to be simpler.

I didn't want to support multiple index files that way, because it would repeat the code to join paths unnecessarily, and make it harder to support user defined implied index file substitution in the future.

**Comment by xuset** [comment]
Everything looks good to me!

**Pull request merged**